src/ring_theory/discriminant.lean

```diff
@@ -0,0 +1,221 @@
+    fintype.sum_congr _ _ (λ (i : ι), @bilin_form.sum_right _ _ _ _ _ _ _ _
+    (b i * P.map (algebra_map A B) i α) (λ (y : ι), b y * P.map (algebra_map A B) y β)), sum_comm],
+  congr, ext x,
+  rw [mul_apply, sum_mul],
+  congr, ext y,
+  rw [map_apply, trace_form_apply, mul_comm (b y), ← smul_def],
+  simp only [id.smul_eq_mul, ring_hom.id_apply, map_apply, transpose_apply, linear_map.map_smulₛₗ,
+    trace_form_apply, algebra.smul_mul_assoc],
+  rw [mul_comm (b x), ← smul_def],
+  ring_nf,
+  simp,
+end
+
+lemma trace_matrix_of_matrix_mul_vec [fintype ι] (b : ι → B) (P : matrix ι ι A) :
+  trace_matrix A ((P.map (algebra_map A B)).mul_vec b) = P ⬝ (trace_matrix A b) ⬝ Pᵀ :=
+begin
+  refine add_equiv.injective transpose_add_equiv _,
+  rw [transpose_add_equiv_apply, transpose_add_equiv_apply, ← vec_mul_transpose,
+    ← transpose_map, trace_matrix_of_matrix_vec_mul, transpose_transpose, transpose_mul,
+    transpose_transpose, transpose_mul]
+end
+
+lemma trace_matrix_of_basis [decidable_eq ι] [fintype ι] (b : basis ι A B) :
+  trace_matrix A b = bilin_form.to_matrix b (trace_form A B) :=
+begin
+  ext i j,
+  rw [trace_matrix, trace_form_apply, trace_form_to_matrix]
+end
+
+variable (A)
+
+/-- `embeddings_matrix A C b : matrix ι (B →ₐ[A] C) C` is the matrix whose `(i, σ)` coefficient is
+  `σ (b i)`. It is mostly useful for fields when `fintype.card ι = finrank A B` and `C` is
+  algebraically closed. -/
+def embeddings_matrix (b : ι → B) : matrix ι (B →ₐ[A] C) C := (λ i (σ : B →ₐ[A] C), σ (b i))
+
+/-- `embeddings_matrix_reindex A C b e : matrix ι ι C` is the matrix whose `(i, j)` coefficient
+  is `σⱼ (b i)`, where `σⱼ : B →ₐ[A] C` is the embedding corresponding to `j : ι` given by a
+  bijection `e : ι ≃ (B →ₐ[A] C)`. It is mostly useful for fields and `C` is algebraically closed.
+  In this case, in presnce of `h : fintype.card ι = finrank A B`, one can take
+  `e := equiv_of_card_eq ((alg_hom.card A B C).trans h.symm)`. -/
+def embeddings_matrix_reindex (b : ι → B) (e : ι ≃ (B →ₐ[A] C)) :=
+reindex (equiv.refl ι) e.symm (embeddings_matrix A C b)
+
+variable {A}
+
+lemma embeddings_matrix_reindex_eq_vandermonde (pb : power_basis A B)
+  (e : fin pb.dim ≃ (B →ₐ[A] C)) :
+  embeddings_matrix_reindex A C pb.basis e = (vandermonde (λ i, e i pb.gen))ᵀ :=
+by { ext i j, simp [embeddings_matrix_reindex, embeddings_matrix] }
+
+end matrix
+
+section discriminant
+
+/-- Given an `A`-algebra `B` and `b`, an `ι`-indexed family of elements of `B`, we define
+`discriminant A ι b` as the determinant of `trace_matrix A ι b`. -/
+noncomputable
+def discriminant (A : Type u) {B : Type v} [comm_ring A] [comm_ring B] [algebra A B] [fintype ι]
+  (b : ι → B) := by { classical, exact (trace_matrix A b).det }
+
+lemma discriminant_def [decidable_eq ι] [fintype ι] (b : ι → B) :
+  discriminant A b = (trace_matrix A b).det := by convert rfl
+
+namespace discriminant
```

Comment:
Putting this in the `discriminant` namespace seems slightly unusual; we don't put lemmas about `matrix.det` under a `det` namespace.